Allow Perl_setlocale to work fully on platforms with positional LC_ALL notation

khwilliamson opened the pull request:
This series of commits  effectively does is to change Perl_setlocale() to return its value in the native format which the libc functions are expecting.

This differs from what it used to return only on platforms which use positional notation and only for LC_ALL when not all categories are set  to the same locale.

khwilliamson committed:
locale.c: Remove no longer used macro

khwilliamson committed:
Create S_native_querylocale_i() and use it

This new function differs from the already-existing plain
querylocale_i() in that it returns in the platform's native format,
instead of the internal=to-perl one.

The internal one is used generally so that code doesn't have to cope
with multiple possible formats.  However, the format of the new locale
in Perl_setlocale() is going to be in native format.  We effectively
translate it into our internal one at the input edge, and that is used
thereafter.

But until this commit, the translation back to native format at the
output edge was incomplete.

This mostly worked because native format differs from locale.c
internal format in just two ways:

One is the locale for LC_NUMERIC.  perl keeps it generally in the C
locale, except for brief intervals which higher level code specifies,
when the real locale is swapped in.  (Actually, this isn't quite true.
If the real locale is indistinguishable from C as far as LC_NUMERIC
goes, perl is happy to use it rather than C, so as to save swapping.)
locale.c had the code in it to translate the internal format back to
native, so it worked for this case.

The other is LC_ALL when not all categories are set to the same locale.
Windows and Linux use 'name=value;' pairs notation, while things derived
from BSD (and others) use a positional notation in which only the values
are given, and the system knows which category a given value is for from
its position in the string.  Perl worked fine for the name=value pairs
notation, because that is the same as its internal one, so no
translation got done, but until this commit, there were issues on
positional platforms.  This seldom got in the way since most people, if
they set the locale at all, will just set LC_ALL to some single 'foo'.

What this commit effectively does is change Perl_setlocale() to return
the value in the native format which the libc functions are expecting.
This differs from what it used to return only on platforms which use the
positional notation and only for LC_ALL when not all categories are set
to the same locale.

The new function subsumes much of the work previously done in
Perl_setlocale(), and it is able to simplify some of that work.

khwilliamson committed:
S_calculate_LC_ALL_string: Add parameter

This parameter tells whether the return value is desired or not,
presenting some optimization possibilities.

khwilliamson committed:
new_LC_ALL: Make parameter meaningful

Prior to this commit, the parameter to this function was unused, and
existed only so that it could be in a table of function pointers all
having the same signature.

The parameter in the rest of them is the new locale that is being
changed to.  It was assumed that this function would just query the
existing locale, which should have been already set up by the time this
function gets called.

But that assumption will be made wrong shortly.  There will be a segment
of code in Perl_setlocale() where things aren't fully updated when this
function gets called.

Therefore change to pay attention to the parameter, which has always
been the correct value anyway.

To be more precise, before this commit, this function called
querylocale() which is expecting things to already be set up; and things
won't necessarily be fully set up, so avoid that call by using the
parameter.

tonycoz commented on locale.c:
```diff
@@ -2683,8 +2692,16 @@ S_calculate_LC_ALL_string(pTHX_ const char ** category_locales_list,
-     * of LC_ALL.
+     * The caller sets 'returning' to
+     *      WANT_MORTAL_TEMP        the function returns the calculated string
+     *                              as a mmortalized temporary, so the caller
```
double m "mmortalized"

Not new: "mortal" implies an SV to me, and a mortal SV and a SAVEFREEPV() PV may be released at different times (SAVEFREESV() in perlguts talks about that).

Unfortunately we don't seem to have a short-hand name for this.

khwilliamson replied:
How about WANT_TEMP_PV

tonycoz replied:
That works.

tonycoz commented on locale.c:
```diff
@@ -2854,13 +2865,25 @@ S_calculate_LC_ALL_string(pTHX_ const char ** category_locales_list,
+                Renew(PL_setlocale_buf, PL_setlocale_bufsize, char);
+            }
+
+            writable_alias = (char *) PL_setlocale_buf;
```
I think the bug here is that `PL_setlocale_buf` is `const  char *`.

save_to_buffer() gets away with no warning because Copy() and Renew() brokenly cast away the const.

khwilliamson replied:
I've been looking at the trees here and not the forest.  This doesn't lead to incorrect results.  The goal is to be able to return in a variable that XS code shouldn't bet able to write to, but we have to be able to initialize and manipulate the string before returning.  What do you suggest?

tonycoz replied:
They can modify it in the same way you are, by using Copy(), or assigning a newly allocated pointer.

Or by casting as you do.

You can't protect it, I think we're better off not getting any more const unsafe.

khwilliamson commented:
On 8/27/23 18:33, Tony Cook wrote:
> ***@***.**** commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In locale.c 
> <https://github.com/Perl/perl5/pull/21419#discussion_r1306765065>:
> 
>>  
>           /* In all cases here, there's nothing we create that needs to be freed,
>            * so leave 'free_if_void_return' set to the default 'false'. */
>       }
>       else {  /* Here, not all categories have the same locale */
> -        char * writable_alias;  /* XXX This name and indentation will make
> -                                   sense in a few commits */
> +        char * writable_alias;
> +
> +        /* If returning to PL_setlocale_buf, set up to write directly to it,
> +         * being sure it is resized to be large enough */
> +        if (returning == WANT_PL_setlocale_buf) {
> +            if (total_len > PL_setlocale_bufsize) {
> +                PL_setlocale_bufsize = total_len;
> +                Renew(PL_setlocale_buf, PL_setlocale_bufsize, char);
> +            }
> +
> +            writable_alias = (char *) PL_setlocale_buf;
> 
> They can modify it in the same way you are, by using Copy(), or 
> assigning a newly allocated pointer.
> 
> Or by casting as you do.
> 
> You can't protect it, I think we're better off not getting any more 
> const unsafe.

Please explain further.

Sure the protections are easily worked around, but the point is to lower 
the odds of it being inadvertently modified.




> 
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/Perl/perl5/pull/21419#discussion_r1306765065>, or 
> unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAA2DH6QS6UQH73LOG2J4YDXXPRNLANCNFSM6AAAAAA322PBNM>.
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>
>

tonycoz commented:
There's plenty of other state that shouldn't be modified incautiously, we document it and move on.

My ideal would be for perl to be `-Wcast-qual` clean as part of const safeness, and we're far from that, but this only makes that worse.

If you really want to prevent inadvertent access you could hide the fields behind an incomplete struct that's only defined in locale.c, and add APIs for any fields that need read access.

khwilliamson committed:
Make PL_setlocale_buf and PL_langinfo_buf non-const

Having them const required some casting away of const.
Also applies to PL_less_dicey_locale_buf

khwilliamson committed:
Split S_save_to_buffer into two

It creates a new, separately callable, function
'set_save_buffer_min_size'  to be used to make sure the buffer is at
least a given size.

save_to_buffer() now calls that new function instead of doing the
resizing itself.

khwilliamson committed:
locale.c: Get rid of a casting-away-const instance

Previously this was used to defeat constness in populating a string.  A
slight refactoring removes that need.

tonycoz commented on locale.c:
```diff
@@ -3823,21 +3823,61 @@ Perl_warn_problematic_locale()
+    const char * individ_locales[LC_ALL_INDEX_] = { NULL };
+
+    switch (parse_LC_ALL_string(lc_all,
+                                (const char **) &individ_locales,
```
The `(const char **) &individ_locales` can just be `individ_locales` due to the way arrays decay into pointers.
